# Iris Flower Classificaction Model using Keras

Importing Modules

In [29]:
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline

seed = 3
np.random.seed(seed)

Loading Data

In [11]:
#!wget 'http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

df = pd.read_csv('iris.data',header=None)

In [12]:
df.head()

0    1    2    3            4
0  5.1  3.5  1.4  0.2  Iris-setosa
1  4.9  3.0  1.4  0.2  Iris-setosa
2  4.7  3.2  1.3  0.2  Iris-setosa
3  4.6  3.1  1.5  0.2  Iris-setosa
4  5.0  3.6  1.4  0.2  Iris-setosa

In [13]:
df.columns = ['Sepal length','Sepal width','Petal length','Petal width','class']
df.head()

Sepal length  Sepal width  Petal length  Petal width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [16]:
dataset = df.values
X = dataset[:,0:4].astype(float)
y = dataset[:,4]

One hot encoding class column

In [18]:
encoder = LabelEncoder()
encoder.fit(y)
encoded_y = encoder.transform(y)
# convert integers to dummy variables (i.e. one hot encoding) using keras np_utils
dummy_y = np_utils.to_categorical(encoded_y)

Neural Network Model

In [36]:
def base_model():
  model = Sequential()
  model.add(Dense(4,input_dim=4,kernel_initializer='normal',activation='relu'))
  model.add(Dense(3,kernel_initializer='normal',activation='softmax'))
  model.compile(loss= 'categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
  return model


Training and testig the model


In [39]:
#Creating estimator for kfold cross val
estimator = KerasClassifier(build_fn=base_model,epochs=200,batch_size=5,verbose = 0)
# creating k folds of data
kfold = KFold(n_splits=10,shuffle=True,random_state=seed)
# training and testing using kfold Cross validatioln
results = cross_val_score(estimator,X,dummy_y,cv=kfold)
print(f"Accuracy: {results.mean()*100: .2f} {results.std()*100: .2f} ")

Accuracy:  96.00  4.42 


#### End result

The model Achieved an accuracy of 96.0

and the Standard Deviation of 4.42